services/search/generate_docs_emebeddings.py

```diff
@@ -0,0 +1,129 @@
+import os
+import time
+import glob
+import argparse
+from dotenv import load_dotenv
+from langchain_text_splitters import RecursiveCharacterTextSplitter, MarkdownHeaderTextSplitter
+from pymilvus import FieldSchema, CollectionSchema, DataType, utility, Collection, connections, model
+
+load_dotenv()
+
+def read_md_files(directory):
+    md_files = glob.glob(f"{directory}/**/*.md", recursive=True)
+    docs = []
+    for file in md_files:
+        with open(file, "r", encoding="utf-8") as f:
+            docs.append((file, f.read()))  # Returning a tuple with file name and content
+    return docs
+
+def split_md_by_sections(file_name, content):
+    headers_to_split_on = [
+        ("##", "Header 2"),
+        ("###", "Header 3"),
+    ]
+    markdown_splitter = MarkdownHeaderTextSplitter(headers_to_split_on=headers_to_split_on, strip_headers=False)
+    md_header_splits = markdown_splitter.split_text(content)
+
+    text_splitter = RecursiveCharacterTextSplitter(chunk_size=900, chunk_overlap=256, length_function=len, is_separator_regex=False)
+    splits = text_splitter.split_documents(md_header_splits)
+
+    # Write the sections to disk
+    output_dir = "./tmp/split_sections"
+    os.makedirs(output_dir, exist_ok=True)
+    with open(os.path.join(output_dir, f"{os.path.basename(file_name)}_sections.md"), "w", encoding="utf-8") as out_file:
+        for section in splits:
+            out_file.write(f"## Section Start: " + "\n------\n")
+            out_file.write(section.page_content + "\n------\n")
+
+    return splits
+
+if __name__ == "__main__":
+    # Parse the command line arguments - repo_path and collection_name
+    parser = argparse.ArgumentParser(description="Initialize Milvus with markdown files")  
+    parser.add_argument("repo_path", type=str, help="Path to the repository containing markdown files")
+    parser.add_argument("collection_name", type=str, help="Name of the Milvus collection")
+    args = parser.parse_args()
+
+    openai_key = os.getenv("OPENAI_API_KEY")
+    openai_ef = model.dense.OpenAIEmbeddingFunction(
+        model_name="text-embedding-ada-002", 
+        api_key=openai_key
+    )
+
+   # Read markdown files from the repo
+    docs_path = args.repo_path
```

Comment:
Are we currently generating embedding for the _whole_ docs repo? That's going to generate a lot of noise. I thought were just indexing the job writing guide, or maybe just the docs subpath

Can you think of a good way to config drive this stuff? We take a path to the docs repo as an argument, but then we need to capture an array of paths _within_  the repo to parse.

I don't think we need to change this config file very often, but we do need the tracked paths to be easily discoverable. A config file called `paths.config` with a glob per line - like a gitignore - might be a good solution here? The file would need to support comments

And as mentioned elsewhere, I think we need to handle the adaptor embeddings a bit differently

Reply:
No, we are not generating embeddings for the whole repo, it is just for the jobs folder currently. We are passing the path to embed as `tmp/repo/docs/jobs`. This works well for embedding a single folder. However, as we plan to embed more documentation in the future, I might need to rethink this approach.

One idea I had while working on this was to maintain a list of folder names that we need to embed, which can be adjusted as needed. For example: `const path = [ "jobs", "adaptors"]`.
What are your thoughts on this approach? I’ll implement the method you suggested today, and then we can compare both approaches to see which one is better in the long run.

About handling the chunking part, I have been working on exploring different methods. One of them is [semantic chunking](https://medium.com/the-ai-forum/semantic-chunking-for-rag-f4733025d5f5) which creates chunks based on similarity between the sentences. This seems to be quite effective. We can use this to chunk the job docs. 
For the adaptor docs we can simply use header splitters and add metadata. I will give updates after I have implemented and tested this.

Apart from this, I have been trying to implement a hybrid approach to retrieve relevant chunks. Milvus uses three search methods:

- Dense Retrieval: Utilizes semantic context to understand the meaning behind queries.
- Sparse Retrieval: Emphasizes keyword matching to find results based on specific terms, equivalent to full-text search.
- Hybrid Retrieval: Combines both Dense and Sparse approaches, capturing the full context and specific keywords for comprehensive search results.

We can read more about this [here](https://milvus.io/docs/hybrid_search_with_milvus.md)

These changes should help improve our embeddings and search results. I'll push the changes after I’ve tested them along with the other updates.